node/core/av-store/src/lib.rs

```diff
@@ -546,601 +546,588 @@ where
 
-	Ok(rx.await??)
-}
+			meta.state = match meta.state {
+				State::Unavailable(at) => {
+					let at_d: Duration = at.into();
+					let prune_at = at_d + pruning_config.keep_unavailable_for;
+					delete_pruning_key(db_transaction, prune_at, &candidate_hash);
 
-#[tracing::instrument(level = "trace", skip(subsystem, ctx), fields(subsystem = LOG_TARGET))]
-async fn process_message<Context>(
-	subsystem: &mut AvailabilityStoreSubsystem,
-	ctx: &mut Context,
-	msg: AvailabilityStoreMessage,
-) -> Result<(), Error>
-where
-	Context: SubsystemContext<Message=AvailabilityStoreMessage>
-{
-	use AvailabilityStoreMessage::*;
+					State::Unfinalized(at, vec![be_block])
+				}
+				State::Unfinalized(at, mut within) => {
+					if let Err(i) = within.binary_search(&be_block) {
+						within.insert(i, be_block);
+					}
 
-	let _timer = subsystem.metrics.time_process_message();
+					State::Unfinalized(at, within)
+				}
+				State::Finalized(_at) => {
+					// This should never happen as a candidate would have to be included after
+					// finality.
+					return Ok(())
+				}
+			};
 
-	match msg {
-		QueryAvailableData(hash, tx) => {
-			tx.send(available_data(&subsystem.inner, &hash).map(|d| d.data)).map_err(|_| oneshot::Canceled)?;
+			write_unfinalized_block_contains(db_transaction, block.0, &block.1, &candidate_hash);
+			write_meta(db_transaction, &candidate_hash, &meta);
 		}
-		QueryDataAvailability(hash, tx) => {
-			let result = available_data(&subsystem.inner, &hash).is_some();
+	}
 
-			tracing::trace!(
-				target: LOG_TARGET,
-				candidate_hash = ?hash,
-				availability = ?result,
-				"Queried data availability",
-			);
+	Ok(())
+}
 
-			tx.send(result).map_err(|_| oneshot::Canceled)?;
-		}
-		QueryChunk(hash, id, tx) => {
-			tx.send(get_chunk(subsystem, &hash, id)?).map_err(|_| oneshot::Canceled)?;
+async fn process_block_finalized(
+	ctx: &mut impl SubsystemContext,
+	subsystem: &AvailabilityStoreSubsystem,
+	finalized_hash: Hash,
+	finalized_number: BlockNumber,
+) -> Result<(), Error> {
+	let now = subsystem.clock.now()?;
+
+	macro_rules! peek_num {
+		($iter:ident) => {
+			match $iter.peek() {
+				Some((k, _)) => decode_unfinalized_key(&k[..]).ok().map(|(b, _, _)| b),
+				None => None
+			}
 		}
-		QueryChunkAvailability(hash, id, tx) => {
-			let result = get_chunk(subsystem, &hash, id).map(|r| r.is_some());
+	}
 
-			tracing::trace!(
-				target: LOG_TARGET,
-				candidate_hash = ?hash,
-				availability = ?result,
-				"Queried chunk availability",
-			);
+	let mut next_possible_batch = 0;
+	loop {
+		let mut db_transaction = DBTransaction::new();
+		let (start_prefix, end_prefix) = finalized_block_range(finalized_number);
+
+		// We have to do some juggling here of the `iter` to make sure it doesn't cross the `.await` boundary
+		// as it is not `Send`. That is why we create the iterator once within this loop, drop it,
+		// do an asynchronous request, and then instantiate the exact same iterator again.
```

Comment:
hmm, we might make the `iter` send if it causes much trouble
not sure if it's possible/safe to do w/o breaking changes, but with https://github.com/paritytech/parity-common/issues/437 (or expose it only in kvdb-rocksdb) it might

Reply:
someday, maybe ;)